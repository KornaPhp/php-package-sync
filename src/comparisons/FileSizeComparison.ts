import { ComparisonKind } from '../types/FileComparisonResult';
import { Comparison } from './Comparison';

export class FileSizeComparison extends Comparison {
    protected filesize1 = 0;
    protected filesize2 = 0;

    public compare(requiredScore: number | null): Comparison {
        this.filesize1 = this.file?.sizeOnDisk ?? 0;
        this.filesize2 = this.repoFile?.sizeOnDisk ?? 0;

        this.score = this.percentage();
        this.comparisonPassed = !this.meetsRequirement(requiredScore);

        if (!this.passed()) {
            this.createIssue(null);
        }

        return this;
    }

    public getKind(): ComparisonKind {
        return ComparisonKind.ALLOWED_SIZE_DIFFERENCE_EXCEEDED;
    }

    public meetsRequirement(percentage): boolean {
        if (percentage === null) {
            percentage = this.file?.requiredScores?.size ?? 15;
        }

        const pctDecimal = percentage * 0.01;

        return this.difference() > Math.round(this.filesize1 * pctDecimal);
    }

    public prettyScore(): string {
        if (this.score === null) {
            return '';
        }

        return (this.score?.toFixed(2) ?? '0.0') + '%';
    }

    public difference() {
        return this.filesize1 - this.filesize2;
    }

    public percentage(): number {
        if (this.difference() === 0) {
            return 0;
        }

        return (this.difference() / this.filesize1) * 100;
    }
}
