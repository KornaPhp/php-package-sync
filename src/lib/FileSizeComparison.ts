/* eslint-disable no-unused-vars */

export class FileSizeComparison {
    static create(filesize1: number, filesize2: number) {
        return new FileSizeComparison(filesize1, filesize2);
    }

    constructor(public filesize1: number, public filesize2: number) {
        //
    }

    get difference(): number {
        return this.filesize1 - this.filesize2;
    }

    get percentDifference(): number {
        const diff = this.difference;

        return diff === 0 ? 0 : (diff / this.filesize1) * 100;
    }

    public differByPercentage(percentage = 15): boolean {
        const pctDecimal = percentage * 0.01;

        return this.difference > Math.round(this.filesize1 * pctDecimal);
    }

    public percentDifferenceForDisplay(padStart = 0, padEnd = 0): string {
        return (this.percentDifference.toFixed(2) + '%').padStart(padStart)
            .padEnd(padEnd);
    }
}

export function compareFileSizes(filesize1: number, filesize2: number): FileSizeComparison {
    return FileSizeComparison.create(filesize1, filesize2);
}
