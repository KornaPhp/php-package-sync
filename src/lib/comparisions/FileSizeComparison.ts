/* eslint-disable no-unused-vars */

export class FileSizeComparison {
    static create(filesize1: number, filesize2: number) {
        return new FileSizeComparison(filesize1, filesize2);
    }

    constructor(public filesize1: number, public filesize2: number) {
        //
    }

    get difference(): number {
        return this.filesize1 - this.filesize2;
    }

    get percentage(): number {
        if (this.difference === 0) {
            return 0;
        }

        return (this.difference / this.filesize1) * 100;
    }

    public meetsRequirement(percentage = 15): boolean {
        if (this.difference === 0) {
            return false;
        }

        const pctDecimal = percentage * 0.01;

        return this.difference >= Math.round(this.filesize1 * pctDecimal);
    }

    public format(decimals = 2, suffix = '%'): string {
        return this.percentage.toFixed(decimals) + suffix;
    }
}

export function compareFileSizes(filesize1: number, filesize2: number): FileSizeComparison {
    return FileSizeComparison.create(filesize1, filesize2);
}
