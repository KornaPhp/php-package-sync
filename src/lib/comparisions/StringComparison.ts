/* eslint-disable no-unused-vars */

const { compareTwoStrings } = require('string-similarity');

export class StringComparison {
    protected _similarityScore: number;

    constructor(protected _string1: string, protected _string2: string) {
        this._similarityScore = compareTwoStrings(this.string1, this.string2);
    }

    static create(string1: string, string2: string) {
        return new StringComparison(string1, string2);
    }

    get similarityScore() {
        return this._similarityScore;
    }

    get string1() {
        return this._string1;
    }

    get string2() {
        return this._string2;
    }

    public meetsRequirement(minimum: number) {
        return this.similarityScore >= minimum;
    }
}
