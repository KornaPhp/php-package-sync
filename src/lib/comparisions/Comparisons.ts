import { FileSizeComparison } from './FileSizeComparison';
import { StringComparison } from './StringComparison';

export class Comparisons {
    public static filesizes(sourceSize: number | undefined, targetSize: number | undefined) {
        return FileSizeComparison.create(sourceSize ?? 0, targetSize ?? 0);
    }

    public static strings(string1: string | undefined, string2: string | undefined) {
        return StringComparison.create(string1 || '', string2 || '');
    }
}
