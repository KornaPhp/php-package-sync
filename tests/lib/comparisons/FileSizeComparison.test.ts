/* eslint-disable no-undef */
import { FileSizeComparison } from '../../../src/lib/comparisions/FileSizeComparison';

it('compares two file sizes', () => {
    expect(FileSizeComparison.create(1000, 2000))
        .toMatchSnapshot();
    expect(FileSizeComparison.create(2000, 1000))
        .toMatchSnapshot();
    expect(FileSizeComparison.create(0, 50))
        .toMatchSnapshot();
    expect(FileSizeComparison.create(100, 100))
        .toMatchSnapshot();
    expect(FileSizeComparison.create(0, 0))
        .toMatchSnapshot();
});

it('checks if a comparison meets requirements', () => {
    expect(FileSizeComparison.create(100, 50)
        .meetsRequirement(20))
        .toBeTruthy();
    expect(FileSizeComparison.create(100, 200)
        .meetsRequirement(75))
        .toBeFalsy();
});

it('returns the correct values for properties', () => {
    expect(FileSizeComparison.create(10, 20).filesize1)
        .toBe(10);
    expect(FileSizeComparison.create(10, 20).filesize2)
        .toBe(20);
    expect(FileSizeComparison.create(20, 10).difference)
        .toBe(10);
});

it('formats the percent difference for display', () => {
    expect(FileSizeComparison.create(100, 20)
        .format(0))
        .toBe('80%');
    expect(FileSizeComparison.create(100, 20)
        .format(1))
        .toBe('80.0%');
});
