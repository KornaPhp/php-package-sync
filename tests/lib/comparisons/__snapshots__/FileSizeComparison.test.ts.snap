// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compares two file sizes 1`] = `
FileSizeComparison {
  "filesize1": 1000,
  "filesize2": 2000,
}
`;

exports[`compares two file sizes 2`] = `
FileSizeComparison {
  "filesize1": 2000,
  "filesize2": 1000,
}
`;

exports[`compares two file sizes 3`] = `
FileSizeComparison {
  "filesize1": 0,
  "filesize2": 50,
}
`;

exports[`compares two file sizes 4`] = `
FileSizeComparison {
  "filesize1": 100,
  "filesize2": 100,
}
`;

exports[`compares two file sizes 5`] = `
FileSizeComparison {
  "filesize1": 0,
  "filesize2": 0,
}
`;
